File: Questions/TwoSum/Optimised/TwoSum.java
package Questions.TwoSum.Optimised;

import java.util.*;

public class TwoSum {
    public static void main(String[] args) {
        int arr[] = { 2, 7, 11, 15 };
        int target = 9;
        int[] result = InnerTwoSum(arr, target);
        System.out.println(Arrays.toString(result));
    }

    static int[] InnerTwoSum(int arr[], int target) { // arr = {2,7,11,15} , target = 9
        HashMap<Integer, Integer> TempMap = new HashMap<>();
        for (int i = 0; i < arr.length; i++) {
            int complement = target - arr[i];
            if (TempMap.containsKey(complement)) {
                return new int[] { TempMap.get(complement), i };
            }
            TempMap.put(arr[i], i);
        }
        return new int[] { -1, -1 };
    }
}

// Optimised version

File: Questions/TwoSum/Optimised/TwoSum.py
arr = [2,7,11,15]
target = 9
def TwoSum(arr, target):
    TempMap = {}
    for i in range(len(arr)):
        complement = target - arr[i]
        if complement in TempMap:
            return [TempMap[complement], i]
        TempMap[arr[i]]=i
    return [-1,-1]

print(TwoSum(arr, target))